### ISSUE TITLE
GetCurrentControllerType unreliable on Oculus SDK

### ISSUE THREAD

**Issue by BHSPitMonkey** [thestonefox/VRTK#1712]
### Environment

 * Source of VRTK - **GitHub**
 * Version of VRTK - **6cb308689b6d786ad62f28bb7e34ca8dec1f6018**
 * Version of Unity3D - **Unity 2017.3.0p3**
 * Hardware used - **Oculus**
 * SDK used - **Oculus Utilities**

### Steps to reproduce

1. Use the Oculus SDK
2. Call `VRTK_SDK_Bridge.GetCurrentControllerType(controllerReference);` from the `Start` method of an object enabled via `VRTK_SDKManager.instance.AddBehaviourToToggleOnLoadedSetupChange(this);` (or, from any point in your game later on while both Touch Controllers are at rest on a surface)

### Expected behavior

Returned value is `SDK_BaseController.ControllerType.Oculus_OculusTouch`

### Current behavior

Returned value is `SDK_BaseController.ControllerType.Undefined`

### Additional information

The current SDK_OculusController implementation [calls `OVRInput.GetActiveController()`](https://github.com/thestonefox/VRTK/blob/release/3.3.0-alpha/Assets/VRTK/Source/SDK/Oculus/SDK_OculusController.cs#L91) to try to determine the controller type. The problem with this is that `GetActiveController()` only returns a meaningful value *while at least one controller is moving or being interacted with*.

A more reliable OVR function for what this VRTK function is aiming to do is `OVRInput.GetConnectedControllers()`, which doesn't take into account what is currently happening to the connected controller(s). However, I'm noticing that this function does not yet know the controller type if called immediately after the VRTK Oculus SDK Setup becomes "loaded". (I am attempting to find the controller type in my Controller class's `Start` method, which runs as soon as the SDK Manager finishes loading the SDK courtesy of `AddBehaviourToToggleOnLoadedSetupChange`.)

**Comment by thestonefox** [comment]
You shouldn't be calling stuff in Start.

You should either call stuff in OnEnable which will be called when the SDK has switched or use the ControllerReady event which will know for sure when the controller is ready.

**Comment by BHSPitMonkey** [comment]
That seems reasonable, though even with that guarantee it still seems GetActiveController is wrong and unpredictable for what VRTK is trying to use it for here. I'll be able to check tomorrow if that modification in conjunction with your suggestion solves my problem.

**Comment by thestonefox** [comment]
You're not supposed to make calls to the VRTK_SDK_Bridge within your code, use the relevant helper classes that make the calls on your behalf.

**Comment by BHSPitMonkey** [comment]
>You're not supposed to make calls to the VRTK_SDK_Bridge within your code, use the relevant helper classes that make the calls on your behalf.

Noted. I've replaced the `VRTK_SDK_Bridge.GetCurrentControllerType` call with its DeviceFinder proxy.

> You shouldn't be calling stuff in Start.

> You should either call stuff in OnEnable which will be called when the SDK has switched or use the ControllerReady event which will know for sure when the controller is ready.

Moving my controller script's "what am I" check from Start to OnEnable hasn't had an effect on this problem (though I don't doubt that it's still a good move). I'm still not sure where the proper/public ControllerReady event is exposed, but that also shouldn't matter because my Controller script is being enabled via the SDK Manager's `AddBehaviourToToggleOnLoadedSetupChange` feature, which (based on my understanding of the `SDKManager.FinishSDKSetupLoading` source) checks that the controllers are "ready" before firing `OnLoadedSetupChanged` at all. This is all working great under the SteamVR SDK, but not with OVR.

I still believe in my suspicion that VRTK.SDK_OculusController should be making calls to `OVRInput.GetConnectedControllers()`, not `OVRInput.GetActiveController()`, for the same reasons I gave above. GetActiveController gives you a different answer every frame depending on the motion of the controller, and in fact if you start your game with the Touch controllers resting on a table flat surface you will see that VRTK's `GetCurrentControllerType` type is actually getting `None` all the way up until the point when you first decide to move a controller. GetConnectedControllers, on the other hand, will give you the answer `Touch` almost immediately after launch (in my observations, beginning on frame 2).

All this is to say, I think that

1. VRTK should be calling `OVRInput.GetConnectedControllers` where it is currently calling `OVRInput.GetActiveController`.
2. VRTK is firing the Oculus SDKSetup's `ControllerReady`/`OnLoadedSetupChanged` event too early (before it even knows what kind of controls it's initially dealing with). (I am getting this call on frame 1, and frame 2 is when GetConnectedControllers or even GetActiveController first gives me a correct answer that isn't `None`).

**Note:** I'm using OculusUtilities 1.21.0 in my project at the moment. I can try updating to 1.22.0 (released today), but I'm not seeing anything in their changelog that seems relevant.

**Comment by BHSPitMonkey** [comment]
I managed to get a reliable "ready" callback by doing the following:

```csharp
    void Awake()
    {
        VRTK_SDKManager.instance.AddBehaviourToToggleOnLoadedSetupChange(this);
    }

    // Called when SDK is loaded but potentially before controllers are ready
    private void OnEnable()
    {
        // Register the appropriate Ready event
        switch (hand)
        {
            case SDK_BaseController.ControllerHand.Left:
                VRTK_SDK_Bridge.GetControllerSDK().LeftControllerReady += OnControllerReady;
                break;
            case SDK_BaseController.ControllerHand.Right:
                VRTK_SDK_Bridge.GetControllerSDK().RightControllerReady += OnControllerReady;
                break;
        }
    }

    // Called when controllers are ready
    private void OnControllerReady (object sender, VRTKSDKBaseControllerEventArgs e)
    {
        var controllerType = VRTK_DeviceFinder.GetCurrentControllerType(controllerReference);
        // set up device-specific input events, etc...
    }
```

(I didn't have any luck finding a better/more-public interface to the Ready events than `VRTK_SDK_Bridge.GetControllerSDK()`)

Even with this problem solved, with VRTK unmodified my controllers don't become "ready" (don't know what type they are) until one of them moves. When I replace `GetActiveController` with `GetConnectedControllers` as suggested above, the controllers become ready very early in the scene (though still not as early as I'd expected; it took over 80 frames to fire).

The other issue I see with `GetActiveController` is that it's being used in a VRTK function that tries to be controller-specific (GetControllerType accepts a controllerRef as a parameter), but the data it returns is just the type of whichever connected controller moved or received input last. On the other hand, `GetConnectedControllers` is giving a more stable and holistic view of the player's input configuration.

**Comment by BHSPitMonkey** [comment]
Change has been implemented in 63339816f4f0018980d5c1a40fb0c5d9c2e5dbe2. Thanks!
